spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,238 @@
+		state' = state.SetSwitch(true),
+	}
+
+	/*
+	 * Start the service.
+	 * If it's already started or stopped, will return an error.
+	 */
+	pure def OnStart(s: ReactorState): ReactorState = {
+		// Startup procedures should come here.
+		s.with("started", true)
+	}
+
+	// Code: Switch.OnStart()
+	action start(): bool = all {
+		state.switch,
+		not(state.started) and not(state.stopped),
+		state' = state.OnStart()
+	}
+
+	/*
+	 * Stop the service.
+	 * If it's already stopped, will return an error.
+	 */
+	pure def OnStop(s: ReactorState): ReactorState = {
+		// Shutdown procedures should come here.
+		s.with("stopped", true)
+	}
+
+	// Code: Switch.OnStop()
+	action stop(): bool = all {
+		IsRunning(state),
+		// Either no peer was added or all peers were removed
+		state.peers.size() == 0,
+		state' = state.OnStop()
+	}
+
+	/*
+	 * Return true if the service is running.
+	 */
+	pure def IsRunning(s: ReactorState): bool = {
+		s.started and not(s.stopped)
+	}
+
+	/*
+	 * InitPeer is called by the switch before the peer is started. Use it to
+	 * initialize data for the peer (e.g. peer state).
+	 */
+	pure def InitPeer(s: ReactorState, peer: Peer): (ReactorState, Peer) = {
+		// This method can update the received peer, which is returned.
+		val updatedPeer = peer
+		(s.with("peers", s.peers.union(Set(peer.ID))), updatedPeer)
+	}
+
+	// Code: Switch.addPeer(p Peer): preamble
+	action initPeer(peer: Peer): bool = all {
+		state.switch and state.IsRunning(),
+		not(peer.ID.in(state.peers)),
+		state' = state.InitPeer(peer)._1 // Updated peer not used here
+	}
+
+	/* 
+	 * AddPeer is called by the switch after the peer is added and successfully
+	 * started. Use it to start goroutines communicating with the peer.
+	 */
+	pure def AddPeer(s: ReactorState, peer: Peer): ReactorState = {
+		// This method can be used to start routines to handle the peer.
+		// Below an example of an arbitrary 'ioRoutine' routine.
+		val startedRoutines = Set( {name: "ioRoutine", peer: peer} )
+		s.with("routines", s.routines.union(startedRoutines))
+	}
+
+	// Code: Switch.addPeer(p Peer): conclusion
+	action addPeer(peer: Peer): bool = all {
+		state.switch and state.IsRunning(),
+		peer.ID.in(state.peers), // InitPeer(peer) and not RemovePeer(peer)
+		state.routines.filter(r => r.peer.ID == peer.ID).size() == 0,
+		state' = AddPeer(state, peer)
+	}
+
+	/*
+	 * RemovePeer is called by the switch when the peer is stopped (due to error
+	 * or other reason).
+	 */
+	pure def RemovePeer(s: ReactorState, peer: Peer, reason: str): ReactorState = {
+		// This method should stop routines created by `AddPeer(Peer)`.
+		val stoppedRoutines = s.routines.filter(r => r.peer.ID == peer.ID)
+		s.with("peers", s.peers.exclude(Set(peer.ID)))
+		 .with("routines", s.routines.exclude(stoppedRoutines))
+	}
+
+	// Code: Switch.stopAndRemovePeer(peer Peer, reason interface{})
+	action removePeer(peer: Peer, reason: str): bool = all {
```

Comment:
Do we need to model "reason", does is affect the behavior of the model?

Reply:
Good point.

If I don't find an actually use for it in the next round of this specification, I will get rid of it.